[Bug]: metanode will panic in some case.

**Describe the bug**
```
2023-07-25 09:22:15,748 runtime.go:44: [ERROR] Recovered from panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(0x1649ea0), concrete:(*runtime._type)(0x1695dc0), asserted:(*runtime._type)(0x1695d20), missingMethod:""} (interface conversion: btree.Item is *metanode.TxRollbackInode, not *metanode.TxRollbackDentry)
/root/tmp/cfs/go/src/github.com/cubefs/cubefs/depends/tiglabs/raft/util/runtime.go:38
/root/tmp/cfs/go/src/github.com/cubefs/cubefs/depends/tiglabs/raft/util/runtime.go:28
/usr/local/go/src/runtime/panic.go:965
/usr/local/go/src/runtime/iface.go:261
/usr/local/go/src/runtime/iface.go:271
/root/tmp/cfs/go/src/github.com/cubefs/cubefs/metanode/transaction.go:1187
/root/tmp/cfs/go/src/github.com/cubefs/cubefs/metanode/partition_fsmop_inode.go:230
/root/tmp/cfs/go/src/github.com/cubefs/cubefs/metanode/partition_fsmop_inode.go:236
```

```
panic: interface conversion: interface {} is nil, not uint8

goroutine 762150 [running]:
github.com/cubefs/cubefs/metanode.(*TransactionManager).commitTx(0xc000017040, 0xc00e90ea20, 0xc, 0xc00071c200, 0xfb0a, 0xc0031f53d8, 0xc01e52a601)
        /root/tmp/cfs/go/src/github.com/cubefs/cubefs/metanode/transaction.go:814 +0x61c
github.com/cubefs/cubefs/metanode.(*metaPartition).TxCommit(0xc00072a0f0, 0xc01e52a600, 0xc013051320, 0x1d949f0, 0xc00072a0f0)
        /root/tmp/cfs/go/src/github.com/cubefs/cubefs/metanode/partition_op_transaction.go:220 +0x65
github.com/cubefs/cubefs/metanode.(*metadataManager).opTxCommit(0xc00061c090, 0x1d893c8, 0xc004fb5db8, 0xc013051320, 0xc00c3e8030
```

**Environment**
develop-3.3.0

--- a/metanode/manager_op.go
+++ b/metanode/manager_op.go
@@ -680,7 +680,7 @@ func (m *metadataManager) opTxMetaUnlinkInode(conn net.Conn, p *Packet, remoteAd
 	}
 	err = mp.TxUnlinkInode(req, p)
 	m.respondToClient(conn, p)
-	log.LogDebugf("%s [opDeleteInode] req: %d - %v, resp: %v, body: %s",
+	log.LogDebugf("%s [opTxMetaUnlinkInode] req: %d - %v, resp: %v, body: %s",
 		remoteAddr, p.GetReqID(), req, p.GetResultMsg(), p.Data)
 	return
 }

--- a/metanode/partition_fsmop_inode.go
+++ b/metanode/partition_fsmop_inode.go
@@ -215,6 +215,10 @@ func (mp *metaPartition) fsmTxUnlinkInode(txIno *TxInode) (resp *InodeResponse) 
 	resp.Status = mp.txProcessor.txResource.addTxRollbackInode(rbInode)
 	if resp.Status == proto.OpExistErr {
 		resp.Status = proto.OpOk
+		item := mp.inodeTree.Get(txIno.Inode)
+		if item != nil {
+			resp.Msg = item.(*Inode)
+		}
 		return
 	}
 	if resp.Status != proto.OpOk {

--- a/metanode/partition_fsmop_transaction.go
+++ b/metanode/partition_fsmop_transaction.go
@@ -160,3 +160,21 @@ func (mp *metaPartition) dentryInTx(parIno uint64, name string) uint8 {
 	}
 	return proto.OpOk
 }
+
+func (mp *metaPartition) txInodeInRb(inode uint64, newTxId string) bool {
+
+	rbIno := mp.txProcessor.txResource.getTxRbInode(inode)
+	if rbIno != nil && rbIno.txInodeInfo.TxID == newTxId {
+		return true
+	}
+
+	return false
+}
+
+func (mp *metaPartition) txDentryInRb(parIno uint64, name, newTxId string) bool {
+	inTx, txId := mp.txProcessor.txResource.isDentryInTransction(&Dentry{
+		ParentId: parIno,
+		Name:     name,
+	})
+	return inTx && txId == newTxId
+}

--- a/metanode/partition_op_dentry.go
+++ b/metanode/partition_op_dentry.go
@@ -172,9 +172,26 @@ func (mp *metaPartition) TxDeleteDentry(req *proto.TxDeleteDentryRequest, p *Pac
 		Name:     req.Name,
 	}
 
+	defer func() {
+		if p.ResultCode == proto.OpOk {
+			var reply []byte
+			resp := &proto.TxDeleteDentryResponse{
+				Inode: req.Ino,
+			}
+			reply, err = json.Marshal(resp)
+			p.PacketOkWithBody(reply)
+		}
+	}()
+
 	dentry, status := mp.getDentry(den)
 	if status != proto.OpOk {
-		err = fmt.Errorf("dentry[%v] not exists", dentry)
+		if mp.txDentryInRb(req.ParentID, req.Name, req.TxInfo.TxID) {
+			p.ResultCode = proto.OpOk
+			log.LogWarnf("TxDeleteDentry: dentry is already been deleted before, req %v", req)
+			return
+		}
+
+		err = fmt.Errorf("dentry[%v] not exists", den)
 		log.LogWarn(err)
 		p.PacketErrorWithBody(status, []byte(err.Error()))
 		return
@@ -208,14 +225,6 @@ func (mp *metaPartition) TxDeleteDentry(req *proto.TxDeleteDentryRequest, p *Pac
 
 	retMsg := r.(*DentryResponse)
 	p.ResultCode = retMsg.Status
-	if p.ResultCode == proto.OpOk {
-		var reply []byte
-		resp := &proto.TxDeleteDentryResponse{
-			Inode: dentry.Inode,
-		}
-		reply, err = json.Marshal(resp)
-		p.PacketOkWithBody(reply)
-	}
 	return
 }
 
@@ -321,14 +330,30 @@ func (mp *metaPartition) TxUpdateDentry(req *proto.TxUpdateDentryRequest, p *Pac
 	}
 
 	txInfo := req.TxInfo.GetCopy()
+
+	defer func() {
+		if p.ResultCode == proto.OpOk {
+			var reply []byte
+			m := &proto.TxUpdateDentryResponse{
+				Inode: req.OldIno,
+			}
+			reply, _ = json.Marshal(m)
+			p.PacketOkWithBody(reply)
+		}
+	}()
+
 	newDentry := &Dentry{
 		ParentId: req.ParentID,
 		Name:     req.Name,
 		Inode:    req.Inode,
 	}
-
 	oldDentry, status := mp.getDentry(newDentry)
 	if status != proto.OpOk {
+		if mp.txDentryInRb(req.ParentID, req.Name, req.TxInfo.TxID) {
+			p.ResultCode = proto.OpOk
+			log.LogWarnf("TxDeleteDentry: dentry is already been deleted before, req %v", req)
+			return
+		}
 		err = fmt.Errorf("oldDentry[%v] not exists", oldDentry)
 		p.PacketErrorWithBody(status, []byte(err.Error()))
 		return
@@ -358,14 +383,6 @@ func (mp *metaPartition) TxUpdateDentry(req *proto.TxUpdateDentryRequest, p *Pac
 
 	msg := resp.(*DentryResponse)
 	p.ResultCode = msg.Status
-	if msg.Status == proto.OpOk {
-		var reply []byte
-		m := &proto.TxUpdateDentryResponse{
-			Inode: oldDentry.Inode,
-		}
-		reply, _ = json.Marshal(m)
-		p.PacketOkWithBody(reply)
-	}
 	return
 }
 

--- a/metanode/partition_op_inode.go
+++ b/metanode/partition_op_inode.go
@@ -184,14 +184,45 @@ func (mp *metaPartition) QuotaCreateInode(req *proto.QuotaCreateInodeRequest, p 
 func (mp *metaPartition) TxUnlinkInode(req *proto.TxUnlinkInodeRequest, p *Packet) (err error) {
 
 	txInfo := req.TxInfo.GetCopy()
+	var status uint8
+	var respIno *Inode
+
+	defer func() {
+		var reply []byte
+		if status == proto.OpOk {
+			resp := &proto.TxUnlinkInodeResponse{
+				Info: &proto.InodeInfo{},
+			}
+
+			replyInfo(resp.Info, respIno, make([]uint32, 0))
+			if reply, err = json.Marshal(resp); err != nil {
+				status = proto.OpErr
+				reply = []byte(err.Error())
+			}
+		}
+		p.PacketErrorWithBody(status, reply)
+	}()
+
 	ino := NewInode(req.Inode, 0)
 	inoResp := mp.getInode(ino)
 	if inoResp.Status != proto.OpOk {
+		if rbIno := mp.txInodeInRb(req.Inode, req.TxInfo.TxID); rbIno {
+			status = proto.OpOk
+			item := mp.inodeTree.Get(NewInode(req.Inode, 0))
+			if item != nil {
+				respIno = item.(*Inode)
+			}
+
+			log.LogWarnf("TxUnlinkInode: inode is already unlink before, req %v, rbIno %s", req, respIno.String())
+			return nil
+		}
+
 		err = fmt.Errorf("ino[%v] not exists", ino.Inode)
 		p.PacketErrorWithBody(inoResp.Status, []byte(err.Error()))
 		return
 	}
 
+	respIno = inoResp.Msg
 	ti := &TxInode{
 		Inode:  inoResp.Msg,
 		TxInfo: txInfo,
@@ -210,25 +241,10 @@ func (mp *metaPartition) TxUnlinkInode(req *proto.TxUnlinkInodeRequest, p *Packe
 	}
 
 	msg := r.(*InodeResponse)
-	status := msg.Status
-	var reply []byte
-	if status == proto.OpOk {
-		resp := &proto.TxUnlinkInodeResponse{
-			Info: &proto.InodeInfo{},
-		}
-
-		ino := ti.Inode
-		if msg.Msg != nil {
-			ino = msg.Msg
-		}
-
-		replyInfo(resp.Info, ino, make([]uint32, 0))
-		if reply, err = json.Marshal(resp); err != nil {
-			status = proto.OpErr
-			reply = []byte(err.Error())
-		}
+	status = msg.Status
+	if msg.Msg != nil {
+		respIno = msg.Msg
 	}
-	p.PacketErrorWithBody(status, reply)
 	return
 }
 

--- a/metanode/transaction.go
+++ b/metanode/transaction.go
@@ -811,12 +811,12 @@ func (tm *TransactionManager) commitTx(txId string, skipSetStat bool) (status ui
 	}
 
 	resp, err := tm.txProcessor.mp.submit(opFSMTxCommit, val)
-	status = resp.(uint8)
-	if err != nil || status != proto.OpOk {
-		log.LogWarnf("commitTx: commit transaction[%v]  failed, err[%v]", txId, err)
+	if err != nil {
+		log.LogWarnf("commitTx: commit transaction[%v] failed, err[%v]", txId, err)
 		return proto.OpTxCommitErr, err
 	}
 
+	status = resp.(uint8)
 	log.LogDebugf("commitTx: tx[%v] is commited successfully", txId)
 
 	return
@@ -932,13 +932,14 @@ func (tm *TransactionManager) rollbackTx(txId string, skipSetStat bool) (status 
 	}
 
 	resp, err := tm.txProcessor.mp.submit(opFSMTxRollback, val)
-	status = resp.(uint8)
-	if err != nil || status != proto.OpOk {
+
+	if err != nil {
 		log.LogWarnf("commitTx: rollback transaction[%v]  failed, err[%v]", txId, err)
 		return proto.OpTxCommitErr, err
 	}
 
-	log.LogDebugf("commitTx: tx[%v] is rollback successfully", txId)
+	status = resp.(uint8)
+	log.LogDebugf("commitTx: tx[%v] is rollback successfully, msg %s", txId, proto.GetStatusStr(status))
 
 	return
 }
@@ -1184,13 +1185,13 @@ func (tr *TransactionResource) deleteTxRollbackInode(ino uint64, txId string) (s
 		return proto.OpTxRbInodeNotExistErr
 	}
 
-	if item.(*TxRollbackDentry).txDentryInfo.TxID != txId {
+	if item.(*TxRollbackInode).txInodeInfo.TxID != txId {
 		log.LogWarnf("deleteTxRollbackInode: rollback dentry is already been update by other, txId %s, item %v",
 			txId, item)
 		return proto.OpTxRbDentryNotExistErr
 	}
 
-	tr.txRbDentryTree.Delete(item)
+	tr.txRbInodeTree.Delete(item)
 	return proto.OpOk
 }
 

--- a/sdk/meta/api.go
+++ b/sdk/meta/api.go
@@ -141,15 +141,19 @@ func (mw *MetaWrapper) txCreate_ll(parentID uint64, name string, mode, uid, gid 
 		return nil, syscall.ENOENT
 	}
 
-	quotaInfos, err := mw.getInodeQuota(parentMP, parentID)
-	if err != nil {
-		log.LogErrorf("Create_ll: get parent quota fail, parentID(%v) err(%v)", parentID, err)
-		return nil, syscall.ENOENT
-	}
 	var quotaIds []uint32
-	for quotaId := range quotaInfos {
-		quotaIds = append(quotaIds, quotaId)
+	if mw.EnableQuota {
+		quotaInfos, err := mw.getInodeQuota(parentMP, parentID)
+		if err != nil {
+			log.LogErrorf("Create_ll: get parent quota fail, parentID(%v) err(%v)", parentID, err)
+			return nil, syscall.ENOENT
+		}
+
+		for quotaId := range quotaInfos {
+			quotaIds = append(quotaIds, quotaId)
+		}
 	}
+
 	rwPartitions = mw.getRWPartitions()
 	length := len(rwPartitions)
 	var tx *Transaction
@@ -1219,17 +1223,6 @@ func (mw *MetaWrapper) txLink(parentID uint64, name string, ino uint64) (info *p
 		return nil, syscall.ENOENT
 	}
 
-	status, info, err = mw.iget(parentMP, parentID)
-	if err != nil || status != statusOK {
-		return nil, statusErrToErrno(status, err)
-	}
-
-	quota := atomic.LoadUint32(&mw.DirChildrenNumLimit)
-	if info.Nlink >= quota {
-		log.LogErrorf("txLink: parent inode's nlink quota reached, parentID(%v)", parentID)
-		return nil, syscall.EDQUOT
-	}
-
 	mp := mw.getPartitionByInode(ino)
 	if mp == nil {
 		log.LogErrorf("txLink: No target inode partition, ino(%v)", ino)
@@ -1261,17 +1254,18 @@ func (mw *MetaWrapper) txLink(parentID uint64, name string, ino uint64) (info *p
 	})
 
 	funcs = append(funcs, func() (int, error) {
-		quotaInfos, err := mw.getInodeQuota(parentMP, parentID)
-		if err != nil {
-			log.LogErrorf("link: get parent quota fail, parentID(%v) err(%v)", parentID, err)
-			return statusError, syscall.ENOENT
-		}
-
 		var quotaIds []uint32
-		for quotaId := range quotaInfos {
-			quotaIds = append(quotaIds, quotaId)
-		}
+		if mw.EnableQuota {
+			quotaInfos, err := mw.getInodeQuota(parentMP, parentID)
+			if err != nil {
+				log.LogErrorf("link: get parent quota fail, parentID(%v) err(%v)", parentID, err)
+				return statusError, syscall.ENOENT
+			}
 
+			for quotaId := range quotaInfos {
+				quotaIds = append(quotaIds, quotaId)
+			}
+		}
 		status, err = mw.txDcreate(tx, parentMP, parentID, name, ino, info.Mode, quotaIds)
 		return status, err
 	})

